cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI static constexpr size_type
+  __size(const _CUDA_VSTD::extents<_IndexType, _Exts...>& __ext) noexcept
+  {
+    size_type __res = 1;
+    for (size_t __i = 0; __i != sizeof...(_Exts); ++__i)
+    {
+      __res *= __ext.extent(__i);
+    }
+    return __res;
+  }
+
+public:
+  //! @addtogroup construction
+  //! @{
+
+  //! @brief Copy-constructs a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! The new async_mdarray has capacity of \p __other.size() which is potentially less than \p __other.capacity().
+  //! @note No memory is allocated if \p __other is empty
+  _CCCL_HIDE_FROM_ABI async_mdarray(const async_mdarray& __other)
+      : __buf_(__other.get_memory_resource(), __other.get_stream(), __other.size())
+      , __mapping_(__other.__mapping_)
+      , __policy_(__other.__policy_)
+  {
+    if (__other.size() != 0)
+    {
+      this->__copy_same(__other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin());
+    }
+  }
+
+  //! @brief Move-constructs a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! The new async_mdarray takes ownership of the allocation of \p __other and resets it.
+  _CCCL_HIDE_FROM_ABI async_mdarray(async_mdarray&& __other) noexcept
+      : __buf_(_CUDA_VSTD::move(__other.__buf_))
+      , __mapping_(_CUDA_VSTD::exchange(__other.__mapping_, mapping_type{}))
+      , __policy_(_CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy))
+  {}
+
+  //! @brief Copy-constructs from a async_mdarray with matching properties
+  //! @param __other The other async_mdarray.
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI explicit async_mdarray(const async_mdarray<_Tp, _OtherProperties...>& __other)
+      : __buf_(__other.get_memory_resource(), __other.get_stream(), __other.size())
+      , __mapping_(__other.__mapping_)
+      , __policy_(__other.__policy_)
+  {
+    if (__other.size() != 0)
+    {
+      this->__copy_cross<const_pointer, __transfer_kind<_OtherProperties...>>(
+        __other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin(), __other.size());
+    }
+  }
+
+  //! @brief Move-constructs from a async_mdarray with matching properties
+  //! @param __other The other async_mdarray.
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI explicit async_mdarray(async_mdarray<_Tp, _OtherProperties...>&& __other) noexcept
+      : __buf_(_CUDA_VSTD::move(__other.__buf_))
+      , __mapping_(_CUDA_VSTD::exchange(__other.__mapping_, mapping_type{}))
+      , __policy_(_CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy))
+  {}
+
+  //! @brief Constructs an empty async_mdarray using an environment
+  //! @param __env The environment providing the needed information
+  //! @note No memory is allocated.
```

Comment:
If "[n]o memory is allocated," then does this mean that the extents are all zero?  If so, then what would happen if all extents are compile-time constants, e.g., `extents<int, 3, 4>`?

Reply:
So currently all constructors are taken from the one dimensional vector case. 

So in this case we should SFINAE that constructor away